File: tts/audio_utils.ts
/**
 * Audio utilities for text-to-speech and audio playback.
 *
 * This module provides functions for:
 * - Generating speech from text using Google Gemini and ElevenLabs TTS APIs
 * - Converting between audio formats (PCM to WAV/MP3)
 * - Playing audio files using ffplay
 */

import * as fs from 'fs';
import * as path from 'path';
import { exec } from 'child_process';
import { promisify } from 'util';
import {GoogleGenAI} from '@google/genai'
import { ElevenLabsClient } from 'elevenlabs';
import * as dotenv from 'dotenv';

// Load environment variables
dotenv.config();

const execAsync = promisify(exec);


/**
 * Save PCM audio data to a WAV file.
 *
 * @param filename - Path where the WAV file will be saved
 * @param pcm - Raw PCM audio data (Buffer)
 * @param channels - Number of audio channels (1=mono, 2=stereo, default: 1)
 * @param rate - Sample rate in Hz (default: 24000)
 * @param sampleWidth - Sample width in bytes (2=16-bit, default: 2)
 */
export function waveFile(
  filename: string,
  pcm: Buffer,
  channels: number = 1,
  rate: number = 24000,
  sampleWidth: number = 2
): void {
  // WAV file header structure
  const header = Buffer.alloc(44);
  const dataSize = pcm.length;
  const fileSize = 36 + dataSize;

  // RIFF chunk descriptor
  header.write('RIFF', 0);
  header.writeUInt32LE(fileSize, 4);
  header.write('WAVE', 8);

  // fmt sub-chunk
  header.write('fmt ', 12);
  header.writeUInt32LE(16, 16); // Subchunk1Size (16 for PCM)
  header.writeUInt16LE(1, 20); // AudioFormat (1 for PCM)
  header.writeUInt16LE(channels, 22);
  header.writeUInt32LE(rate, 24);
  header.writeUInt32LE(rate * channels * sampleWidth, 28); // ByteRate
  header.writeUInt16LE(channels * sampleWidth, 32); // BlockAlign
  header.writeUInt16LE(sampleWidth * 8, 34); // BitsPerSample

  // data sub-chunk
  header.write('data', 36);
  header.writeUInt32LE(dataSize, 40);

  // Write header and PCM data to file
  fs.writeFileSync(filename, Buffer.concat([header, pcm]));
}

/**
 * Convert PCM audio data to MP3 format and save to file.
 *
 * @param pcmData - Raw PCM audio data (Buffer)
 * @param outputFilename - Path to save the MP3 file
 * @param channels - Number of audio channels (default: 1 for mono)
 * @param sampleWidth - Sample width in bytes (default: 2 for 16-bit)
 * @param frameRate - Sample rate in Hz (default: 24000)
 * @returns Path to the created MP3 file
 */
export async function pcmToMp3(
  pcmData: Buffer,
  outputFilename: string,
  channels: number = 1,
  sampleWidth: number = 2,
  frameRate: number = 24000
): Promise<string> {
  try {
    // Create a temporary WAV file from PCM data
    const tempWavFile = `${outputFilename}.temp.wav`;
    waveFile(tempWavFile, pcmData, channels, frameRate, sampleWidth);

    // Convert WAV to MP3 using ffmpeg
    await execAsync(`ffmpeg -i "${tempWavFile}" -codec:a libmp3lame -qscale:a 2 "${outputFilename}" -y`);

    // Remove temporary WAV file
    fs.unlinkSync(tempWavFile);

    return outputFilename;
  } catch (e) {
    const error = e as Error;
    console.error(`Error converting PCM to MP3: ${error.message}`);
    throw error;
  }
}

/**
 * Generate speech from text using Google Gemini TTS API and save as MP3.
 *
 * @param voice - Voice name from Gemini's prebuilt voice options
 * @param text - Text to convert to speech
 * @param filename - Base filename for the output MP3 file (without extension)
 * @returns Path to the generated MP3 file
 * @throws Error if all 3 retry attempts fail
 */
export async function speak(
  voice: string,
  text: string,
  filename: string
): Promise<string> {
  // Retry up to 3 times in case of transient errors
  const addPause = `"${text}", wait 2 seconds after speaking`;

  for (let i = 0; i < 3; i++) {
    try {
      // Initialize Google Gemini client with API key from environment
      const apiKey = process.env.GOOGLE_API_KEY;
      if (!apiKey) {
        throw new Error('GOOGLE_API_KEY environment variable is not set');
      }

      const genAI = new GoogleGenAI({apiKey});

      const response = await genAI.models.generateContent({
        model: "gemini-2.5-flash-preview-tts",
        contents: [{ parts: [{ text:addPause }] }],
        config: {
              responseModalities: ['AUDIO'],
              speechConfig: {
                voiceConfig: {
                    prebuiltVoiceConfig: { voiceName: 'Kore' },
                },
              },
          },
        });

      // Extract PCM audio data from the response
      const data = response.candidates?.[0]?.content?.parts?.[0]?.inlineData?.data;
      if (!data) {
        throw new Error('No audio data received from Gemini API');
      }

      // Convert base64 to Buffer
      const pcmBuffer = Buffer.from(data, 'base64');

      // Convert PCM data to MP3 format and save
      const fileName = `${filename}.mp3`;
      await pcmToMp3(pcmBuffer, fileName);

      return fileName;
    } catch (e) {
      // Log detailed error information for debugging
      const error = e as Error;
      console.error(`Error in speak (attempt ${i + 1}/3): ${error.message}`);
      console.error(`Exception type: ${error.name}`);
      console.error(`Stack trace:`, error.stack);

      if (i === 2) {
        // Last attempt failed
        throw error;
      }
    }
  }

  throw new Error('All retry attempts failed');
}


/**
 * Generate speech from text using ElevenLabs TTS API and save as MP3.
 *
 * @param voice - ElevenLabs voice ID
 * @param text - Text to convert to speech
 * @param filename - Path where the MP3 file will be saved (without extension)
 * @returns Path to the MP3 file, or null on error
 */
export async function speak11(
  voice: string,
  text: string,
  filename: string
): Promise<string | null> {
  text = `"${text}<break time=\"2s\" />`;
  try {
    // Initialize ElevenLabs client with API key
    const apiKey = process.env.ELEVENLABS_API_KEY;
    if (!apiKey) {
      throw new Error('ELEVENLABS_API_KEY environment variable is not set');
    }

    const elevenlabs = new ElevenLabsClient({ apiKey });

    // Convert text to speech using specified voice
    // Output format: MP3 at 44.1kHz sample rate with 128kbps bitrate
    const audio = await elevenlabs.textToSpeech.convert(voice, {
      text,
      output_format: 'mp3_44100_128',
    });

    // Write the audio stream to file
    const outputPath = `${filename}.mp3`;
    const chunks: Buffer[] = [];

    // Collect all chunks from the stream
    for await (const chunk of audio) {
      chunks.push(Buffer.from(chunk));
    }

    // Combine all audio chunks into a single buffer and write to file
    const audioBuffer = Buffer.concat(chunks);
    fs.writeFileSync(outputPath, audioBuffer);

    return outputPath;
  } catch (e) {
    // Log error details and return null to indicate failure
    const error = e as Error;
    console.error(`Error in speak11: ${error.message}`);
    console.error(error.stack);
    return null;
  }
}


/**
 * Play WAV audio file using ffplay.
 *
 * @param wavFile - Path to the WAV file to play
 * @throws Error if ffplay fails to execute
 */
export async function playWave(wavFile: string): Promise<void> {
  try {
    // Run ffplay with:
    // -nodisp: Don't show video display window
    // -autoexit: Exit when playback finishes
    await execAsync(`ffplay -nodisp -autoexit "${wavFile}"`);
  } catch (e) {
    const error = e as any;
    console.error(`Error running ffplay: ${error.message}`);
    console.error(`stderr: ${error.stderr}`);
    throw error;
  }
}

/**
 * Play MP3 audio file using ffplay.
 *
 * @param mp3File - Path to the MP3 file to play
 * @throws Error if ffplay fails to execute
 */
export async function playMp3(mp3File: string): Promise<void> {
  try {
    // Run ffplay with:
    // -nodisp: Don't show video display window
    // -autoexit: Exit when playback finishes
    await execAsync(`ffplay -nodisp -autoexit "${mp3File}"`);
  } catch (e) {
    const error = e as any;
    console.error(`Error running ffplay: ${error.message}`);
    console.error(`stderr: ${error.stderr}`);
    throw error;
  }
}

File: ai-server/internal/tts/audio_utils.go
package tts

import (
	"bytes"
	"context"
	"encoding/binary"
	"fmt"
	"os"
	"os/exec"

	"google.golang.org/genai"
)

// WaveFile saves PCM audio data to a WAV file.
//
// filename: Path where the WAV file will be saved
// pcm: Raw PCM audio data
// channels: Number of audio channels (1=mono, 2=stereo, default: 1)
// rate: Sample rate in Hz (default: 24000)
// sampleWidth: Sample width in bytes (2=16-bit, default: 2)
func WaveFile(filename string, pcm []byte, channels int, rate int, sampleWidth int) error {
	file, err := os.Create(filename)
	if err != nil {
		return fmt.Errorf("failed to create file: %w", err)
	}
	defer file.Close()

	dataSize := uint32(len(pcm))
	fileSize := uint32(36 + dataSize)

	// WAV file header structure
	buf := new(bytes.Buffer)

	// RIFF chunk descriptor
	buf.WriteString("RIFF")
	if err := binary.Write(buf, binary.LittleEndian, fileSize); err != nil {
		return err
	}
	buf.WriteString("WAVE")

	// fmt sub-chunk
	buf.WriteString("fmt ")
	if err := binary.Write(buf, binary.LittleEndian, uint32(16)); err != nil { // Subchunk1Size (16 for PCM)
		return err
	}
	if err := binary.Write(buf, binary.LittleEndian, uint16(1)); err != nil { // AudioFormat (1 for PCM)
		return err
	}
	if err := binary.Write(buf, binary.LittleEndian, uint16(channels)); err != nil {
		return err
	}
	if err := binary.Write(buf, binary.LittleEndian, uint32(rate)); err != nil {
		return err
	}
	// ByteRate = SampleRate * NumChannels * BitsPerSample/8
	if err := binary.Write(buf, binary.LittleEndian, uint32(rate*channels*sampleWidth)); err != nil {
		return err
	}
	// BlockAlign = NumChannels * BitsPerSample/8
	if err := binary.Write(buf, binary.LittleEndian, uint16(channels*sampleWidth)); err != nil {
		return err
	}
	// BitsPerSample
	if err := binary.Write(buf, binary.LittleEndian, uint16(sampleWidth*8)); err != nil {
		return err
	}

	// data sub-chunk
	buf.WriteString("data")
	if err := binary.Write(buf, binary.LittleEndian, dataSize); err != nil {
		return err
	}

	// Write header to file
	if _, err := file.Write(buf.Bytes()); err != nil {
		return fmt.Errorf("failed to write header: %w", err)
	}

	// Write PCM data to file
	if _, err := file.Write(pcm); err != nil {
		return fmt.Errorf("failed to write pcm data: %w", err)
	}

	return nil
}

// PcmToMp3 converts PCM audio data to MP3 format and saves to file.
//
// pcmData: Raw PCM audio data
// outputFilename: Path to save the MP3 file
// channels: Number of audio channels (default: 1 for mono)
// sampleWidth: Sample width in bytes (default: 2 for 16-bit)
// frameRate: Sample rate in Hz (default: 24000)
// Returns Path to the created MP3 file
func PcmToMp3(pcmData []byte, outputFilename string, channels int, sampleWidth int, frameRate int) (string, error) {
	// Create a temporary WAV file from PCM data
	tempWavFile := outputFilename + ".temp.wav"
	if err := WaveFile(tempWavFile, pcmData, channels, frameRate, sampleWidth); err != nil {
		return "", err
	}
	defer os.Remove(tempWavFile)

	// Convert WAV to MP3 using ffmpeg
	// -i: input file
	// -codec:a libmp3lame: audio codec
	// -qscale:a 2: audio quality (variable bitrate)
	// -y: overwrite output file
	cmd := exec.Command("ffmpeg", "-i", tempWavFile, "-codec:a", "libmp3lame", "-qscale:a", "2", outputFilename, "-y")
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("ffmpeg error: %s: %w", string(output), err)
	}

	return outputFilename, nil
}

// Speak generates speech from text using Google Gemini TTS API and saves as MP3.
//
// ctx: Context for the API call
// voice: Voice name from Gemini's prebuilt voice options (e.g., "Kore")
// text: Text to convert to speech
// filename: Base filename for the output MP3 file (without extension)
// Returns Path to the generated MP3 file
func Speak(ctx context.Context, voice string, text string, filename string) (string, error) {
	// Retry up to 3 times in case of transient errors
	addPause := fmt.Sprintf(`"%s", wait 2 seconds after speaking`, text)

	var lastErr error

	apiKey := os.Getenv("GOOGLE_API_KEY")
	if apiKey == "" {
		return "", fmt.Errorf("GOOGLE_API_KEY environment variable is not set")
	}

	client, err := genai.NewClient(ctx, &genai.ClientConfig{APIKey: apiKey})
	if err != nil {
		return "", fmt.Errorf("failed to create genai client: %w", err)
	}

	for i := 0; i < 3; i++ {
		resp, err := client.Models.GenerateContent(ctx, "gemini-2.5-flash-preview-tts",
			genai.Text(addPause),
			&genai.GenerateContentConfig{
				ResponseModalities: []string{"AUDIO"},
				SpeechConfig: &genai.SpeechConfig{
					VoiceConfig: &genai.VoiceConfig{
						PrebuiltVoiceConfig: &genai.PrebuiltVoiceConfig{
							VoiceName: voice,
						},
					},
				},
			},
		)

		if err != nil {
			lastErr = err
			fmt.Printf("Error in speak (attempt %d/3): %v\n", i+1, err)
			continue
		}

		// Extract PCM audio data from the response
		// Structure navigation depends on the specific genai library response format.
		// Assuming standard Candidates -> Content -> Parts -> InlineData
		if len(resp.Candidates) == 0 || resp.Candidates[0].Content == nil || len(resp.Candidates[0].Content.Parts) == 0 {
			lastErr = fmt.Errorf("no content in response")
			fmt.Printf("Error in speak (attempt %d/3): %v\n", i+1, lastErr)
			continue
		}

		part := resp.Candidates[0].Content.Parts[0]

		// The Go library handles InlineData differently than the raw JSON or JS SDK usually.
		// We expect the part to be of type Blob or similar containing the data.
		// In the google.golang.org/genai library, Parts are often interface{}.
		// We need to check if it has InlineData.

		var pcmBuffer []byte

		if part.InlineData != nil {
			pcmBuffer = part.InlineData.Data
		} else {
			lastErr = fmt.Errorf("part does not contain InlineData")
			fmt.Printf("Error in speak (attempt %d/3): %v\n", i+1, lastErr)
			continue
		}

		if len(pcmBuffer) == 0 {
			lastErr = fmt.Errorf("empty audio data received")
			fmt.Printf("Error in speak (attempt %d/3): %v\n", i+1, lastErr)
			continue
		}

		// Convert PCM data to MP3 format and save
		finalFileName := filename + ".mp3"
		if _, err := PcmToMp3(pcmBuffer, finalFileName, 1, 2, 24000); err != nil {
			lastErr = err
			fmt.Printf("Error converting to MP3 (attempt %d/3): %v\n", i+1, err)
			continue
		}

		return finalFileName, nil
	}

	return "", fmt.Errorf("all retry attempts failed: %w", lastErr)
}

// PlayWave plays a WAV audio file using ffplay.
//
// wavFile: Path to the WAV file to play
func PlayWave(wavFile string) error {
	// Run ffplay with:
	// -nodisp: Don't show video display window
	// -autoexit: Exit when playback finishes
	cmd := exec.Command("ffplay", "-nodisp", "-autoexit", wavFile)

	// Capture stderr to report detailed errors if needed, similar to TS
	var stderr bytes.Buffer
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("error running ffplay: %w, stderr: %s", err, stderr.String())
	}
	return nil
}

// PlayMp3 plays an MP3 audio file using ffplay.
//
// mp3File: Path to the MP3 file to play
func PlayMp3(mp3File string) error {
	// Run ffplay with:
	// -nodisp: Don't show video display window
	// -autoexit: Exit when playback finishes
	cmd := exec.Command("ffplay", "-nodisp", "-autoexit", mp3File)

	var stderr bytes.Buffer
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		return fmt.Errorf("error running ffplay: %w, stderr: %s", err, stderr.String())
	}
	return nil
}
